File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int n = mat.size(), m = mat[0].size();
        int xini = n-1, xend = n-1, yini = 0, yend = 0;
        while(xend >= 0){
            vector<int> arr;
            for(int i = xini, j = yini; i <= xend; i++, j++){
                arr.push_back(mat[i][j]);
            }
            sort(begin(arr), end(arr));
            int idx = 0;
            for(int i = xini, j = yini; i <= xend; i++, j++){
                mat[i][j] = arr[idx++];
            }
            if(xini > 0) xini--;
            else {
                yini++;
            }
            if(yend < m-1) yend++;
            else {
                xend--;
            }
        }
        return mat;
    }
};

File: 1329-sort-the-matrix-diagonally/1329-sort-the-matrix-diagonally.java
class Solution {
    public int[][] diagonalSort(int[][] mat) {
        int n = mat.length, m = mat[0].length;
        Map<Integer, PriorityQueue<Integer>> map = new HashMap<>();
        for(int i = 0; i < mat.length; i++){
            for(int j = 0; j < mat[0].length; j++){
                map.
                    computeIfAbsent(i - j, k -> new PriorityQueue<>())
                    .offer(mat[i][j]);
            }
        }
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                mat[i][j] = map.get(i-j).poll();
            }
        }
        return mat;
    }
}